src/sdk/namespace_fs.js

```diff
@@ -2419,15 +2448,35 @@ class NamespaceFS {
             try {
                 upload_params = await this._start_upload(fs_context, undefined, undefined, params, 'w');
 
-                // the delete marker path built from its version info (mtime + ino)
+                let file_path;
                 const stat = await upload_params.target_file.stat(fs_context);
-                delete_marker_version_id = this._get_version_id_by_stat(stat);
-                const file_path = this._get_version_path(params.key, delete_marker_version_id);
+                if (this._is_versioning_enabled()) {
+                    // the delete marker path built from its version info (mtime + ino)
+                    delete_marker_version_id = this._get_version_id_by_stat(stat);
+                    file_path = this._get_version_path(params.key, delete_marker_version_id);
+                } else {
+                    // the delete marker file name would be with a 'null' suffix
+                    delete_marker_version_id = NULL_VERSION_ID;
+                    file_path = this._get_version_path(params.key, delete_marker_version_id);
+                }
 
                 // finish upload part
-                const fs_xattr = await this._assign_versions_to_fs_xattr(fs_context, deleted_version_info,
+                let fs_xattr;
+                if (this._is_versioning_suspended() &&
+                    (deleted_version_info?.version_id_str === NULL_VERSION_ID)) {
+                    // the previous version will be the previous version of the deleted version
+                    const prev_version_info = await this.get_prev_version_info(fs_context, params.key,
```

Comment:
in the design meetings, we decided to exclude this prev linked list maintenance due to concurrent deletions/puts, so maybe let's try to re think about it in a different PR/design